Color.js package

OmarIthawi opened the pull request:
The [color.js packages imported from v2](https://github.com/mathjax/MathJax/blob/master/unpacked/extensions/TeX/color.js)

### Changes
 - [x] Import the package and convert to TypeScript
 - [x] Make the padding and border configurable
 - [x] JSDocs with `@param` and `@return`
 - [x] TSLint
 - [x] Test all the functions

### Larger Review Comments
 - [x] Make sure `\color` works as a switch. See [David's comment](https://github.com/mathjax/mathjax-v3/pull/136#discussion_r224060747).
 - [x] `COLORS` should have a new instance per parser instead of being a global state effectively. See [David's comment](https://github.com/mathjax/mathjax-v3/pull/136#discussion_r224064198).
 - [x] `\textcolor` implementation isn't correct. See [David's comment](https://github.com/mathjax/mathjax-v3/pull/136#discussion_r225008026).
 - [x] Split into multiple files: Constants, Methods, others. [See base options](https://github.com/mathjax/mathjax-v3/blob/master/mathjax3-ts/input/tex/base/BaseConfiguration.ts#L139-L143), and [how it's being used](https://github.com/mathjax/mathjax-v3/blob/master/mathjax3-ts/input/tex/base/BaseMethods.ts#L1568-L1570).

### Not Done
 - [ ] Allow multiple color namespaces color package for CSS, SVG and others.

### Screenshot
![image](https://user-images.githubusercontent.com/645156/52363471-6cee9680-2a4b-11e9-86a4-ca405035c9dd.png)



### Testing TeX
```latex
\definecolor{salty}{named}{#888}

2bc \ { \color{blue} 1yz } \ 2\pi^2  \ldots

\textcolor[gray]{0.75}{ams} \ldots

\textcolor{#f80}{\text{Orange}} \ldots

\colorbox{green}{ams} \ldots 

\fcolorbox{green}{silver}{ Text } \ldots


\color[rgb]{0.5,0.5,1}
H_2O \ldots

\color{salty}
NaCl \ldots
```

zorkow commented:
Great! Our first contributor.

OmarIthawi commented:
@zorkow I've just started importing the plugin. I hope that it's not too late.

OmarIthawi commented:
As a side note, we need a CI setup here!

zorkow commented on mathjax3-ts/input/tex/color/ColorConfiguration.ts:
```diff
@@ -0,0 +1,201 @@
+/**
+ * @fileoverview Configuration file for the color package.
+ *
+ * @author v.sorge@mathjax.org (Volker Sorge)
```
Author name?

OmarIthawi replied:
Done.

zorkow commented on mathjax3-ts/input/tex/color/ColorConfiguration.ts:
```diff
@@ -0,0 +1,201 @@
+
+import {Configuration} from '../Configuration.js';
+import TexParser from '../TexParser.js';
+import {AllTags, NoTags} from '../Tags';
```
Do you need that import? I don't think tags are used in the color extension.

OmarIthawi replied:
Removed it. Thanks!

zorkow commented on mathjax3-ts/input/tex/color/ColorConfiguration.ts:
```diff
@@ -0,0 +1,201 @@
+import {AllTags, NoTags} from '../Tags';
+
+// This is from the 'no undefined' extension.
+function color(parser: TexParser, name: string) {
```
Missing jsdoc.

zorkow replied:
I see that's just a test function.

OmarIthawi replied:
Yup!

zorkow commented on mathjax3-ts/input/tex/color/ColorConfiguration.ts:
```diff
@@ -0,0 +1,201 @@
+    ['YellowOrange',   '#FAA21A']
+]);
+
+/*
```
Start jsdoc with /**

OmarIthawi replied:
Done.

zorkow commented on mathjax3-ts/input/tex/color/ColorConfiguration.ts:
```diff
@@ -0,0 +1,201 @@
+
+/*
+ *  Look up a color based on its model and definition
+ */
```
missing `@param` , `@return`

OmarIthawi replied:
I'll add those later.

OmarIthawi replied:
Done.

zorkow commented:
No worries, I've only just returned from holidays myself.
Currently I am testing using the top level `tests.sh` script and tests in the `tests` directory. We have not yet integrated this into travis but should probably do that.

OmarIthawi commented:
@zorkow I've mostly completed it. If finally ticked and I was able to get this package running!

I've added a screenshot and a couple of TODO items.

I'll get back to it later on. Meanwhile, could you please take a look?

dpvc commented on mathjax3-ts/input/tex/color/ColorConfiguration.ts:
```diff
@@ -0,0 +1,336 @@
+    var cname = parser.GetArgument(name),
+    math = ParseUtil.internalMath(parser, parser.GetArgument(name));
+
+    const node = parser.configuration.nodeFactory.create('node', 'mpadded', [math], {
```
I think `math` is already an array of MathML nodes, not a single one, so you should not need `[math]` here, only `math`.

OmarIthawi replied:
Done.

dpvc commented on mathjax3-ts/input/tex/color/ColorConfiguration.ts:
```diff
@@ -0,0 +1,336 @@
+        cname = parser.GetArgument(name),
+        math = ParseUtil.internalMath(parser, parser.GetArgument(name));
+
+    const node = parser.configuration.nodeFactory.create('node', 'mpadded', [math], {
```
ditto

OmarIthawi replied:
Done.

OmarIthawi commented:
@dpvc Thanks for the detailed review, I'll take look later and get back to you.

OmarIthawi commented on mathjax3-ts/input/tex/color/ColorConfiguration.ts:
```diff
@@ -0,0 +1,336 @@
+        throw new TexError('ModelArg1', 'Color values for the %1 model require 3 numbers', 'rgb');
+    }
+
+    for (var i = 0; i < 3; i++) {
```
Done.

OmarIthawi commented on mathjax3-ts/input/tex/color/ColorConfiguration.ts:
```diff
@@ -0,0 +1,336 @@
+ *  Get an rgb color
+ */
+function get_rgb( rgb: string) {
+    var rgbParts: string[] = rgb.replace(/^\s+/, '').replace(/\s+$/, '').split(/\s*, \s*/); 
```
Done.

OmarIthawi commented on mathjax3-ts/input/tex/color/ColorConfiguration.ts:
```diff
@@ -0,0 +1,336 @@
+ *  Get an RGB color
+ */
+function get_RGB(rgb: string) {
+    var rgbParts : string[] = rgb.replace(/^\s+/,'').replace(/\s+$/,'').split(/\s*,\s*/); 
```
Done.

OmarIthawi commented on mathjax3-ts/input/tex/color/ColorConfiguration.ts:
```diff
@@ -0,0 +1,336 @@
+    if (rgbParts.length !== 3)
+    {throw new TexError('ModelArg1', 'Color values for the %1 model require 3 numbers', 'RGB')}
+    
+    for (var i = 0; i < 3; i++) {
```
Done

OmarIthawi commented on mathjax3-ts/input/tex/color/ColorConfiguration.ts:
```diff
@@ -0,0 +1,336 @@
+    parser.stack.env['color'] = color;
+    const math = parser.ParseArg(name);
+    if (old) {
+      parser.stack.env['color'];  // TODO: Volker, shouldn't we remove this?
```
Thanks! Done.

OmarIthawi commented:
@dpvc @zorkow I've addressed all of the mini issues including TSLint and JSDoc.

I've left a couple undone larger notes for the `{ \color{red} xyz }` bug and the `COLORS` map. They're in a TODO list in the pull request description above (so I don't loose track of 'em).

dpvc commented:
Thanks, this is looking a lot better.  Looking forward to the final fixes.

zorkow commented on mathjax3-ts/input/tex/color/ColorConfiguration.ts:
```diff
@@ -0,0 +1,336 @@
+        model = parser.GetArgument(name),
+        def = parser.GetArgument(name);
+    
+    COLORS.set(cname, getColor(model, def));
```
Effectively, we need a mechanism similar to the map handlers for parse definitions.
They allow to shadow definitions without actually overwriting them. Currently they are initialised creation `TeX` object creation, keeping the definitions local to each instance.
But if we want to have that possibility for custom structures to do something similiar, then we can either 
1. have an interface that these structures have to implement so they get initialised, or alternatively
2. each package configuration can bring it's own init method, that is called when a new `TeX` instance is created. 

The latter is probably easier and a bit more flexible. 
I'll have a think about it in the context of the `newcommand` and `ams` package (`DeclareMathOperator`) after PR #116 is merged.
We probably also need something similar for the `begingroup` package as well.

@OmarIthawi For now, just leave it like this. I am hoping to extend color to xcolor after you are done, which provides multiple named colour mappings. It might be easier to see if we can find a suitable abstraction then.

dpvc replied:
@zorkow, I agree that 2 is probably the right choice, and that it can wait for now.  Just wanted it to be something that we keep in mind; as you point out, there are several other places that it may be important.

OmarIthawi commented:
Thanks @dpvc and @zorkow, I'm continuing working on the remaining bug. I have a question though:

I've hardcoded the padding values, but how I'm a supposed to make them configurable?

```
// TODO: Make this configurable
const ColorConfigs = {
    padding: '5px',
    border: '2px',
};
```

OmarIthawi commented:
@dpvc @zorkow I've addressed all the review comments. Please take a look and let me know what do you think. 

I've also tested all the functions and it seems that they work fine.

dpvc commented on mathjax3-ts/input/tex/color/ColorConfiguration.ts:
```diff
@@ -0,0 +1,376 @@
+    const old = parser.stack.env['color'];
+
+    parser.stack.env['color'] = color;
+    const math = ParseUtil.internalMath(parser, parser.GetArgument(name));
```
Despite its name, `\textcolor` does *not* switch into text mode in LaTeX, so your original was correct here (using `ParseArg()` rather than `internalMath()`).  The contents of `\textcolor` should be processed in math-mode.

OmarIthawi replied:
Thanks @dpvc. There's something I don't understand here, I need to explore more and see how can I address your feedback.

dpvc commented on mathjax3-ts/input/tex/color/ColorConfiguration.ts:
```diff
@@ -0,0 +1,376 @@
+}, ColorMethods);
+
+export const ColorConfiguration = Configuration.create(
+    'color', { handler: { macro: ['color'] } });
```
Why have the other commands been removed?  Don't they have to be here in order to be defined?

OmarIthawi replied:
@dpvc it turns out they don't. There's something I'm not sure about since I put them I get the following warning: `Configuration textcolor not found! Omitted.` for all of them.

Removing those entries fixed the warning while keeping the package functional. Am I missing something @zorkow?

https://github.com/mathjax/mathjax-v3/blob/tex_input_typescript/mathjax3-ts/input/tex/MapHandler.ts#L91-L102

dpvc commented:
I will let @zorkow describe how to do the configuration options.

OmarIthawi commented:
Thanks @dpvc! I'll reach out to @zorkow.

zorkow commented:
Reopening after rebasing.

OmarIthawi commented:
@dpvc I've fixed the `\textcolor` and here's the result:

### Before
![image](https://user-images.githubusercontent.com/645156/47607581-2d62e800-da22-11e8-8b2e-c59be7c666e3.png)

### After
![image](https://user-images.githubusercontent.com/645156/46915218-3038f400-cfb1-11e8-8d07-ed3bb3105556.png)

OmarIthawi commented:
@zorkow @dpvc I've addressed all outstanding review notes and TODOs. I hope that the new update isn't too sloppy for it to get merged.

Please take a look and let me know what do you think.

OmarIthawi commented on mathjax3-ts/input/tex/color/ColorMethods.ts:
```diff
@@ -0,0 +1,298 @@
+    const style = parser.itemFactory.create('style')
+                        .setProperties({styles: { mathcolor: color }});
+    parser.stack.env['color'] = color;
+
```
```suggestion
```

OmarIthawi commented on mathjax3-ts/input/tex/color/ColorMethods.ts:
```diff
@@ -0,0 +1,298 @@
+    const model = parser.GetBrackets(name, '');
+    const colorDef = parser.GetArgument(name);
+    const color = getColor(parser.options, model, colorDef);
+
```
```suggestion
```

OmarIthawi commented:
Thanks @zorkow, all the notes looks good to me, I'll address them soon. Some questions though:

> As for other colour ranges: We should implement that in the integration of the xcolor package.

 - What are color ranges?
 - What I'm I expected to do in this PR? 
 - Is `xcolor` package in V3 now?

zorkow commented:
> * What are color ranges?

Colour ranges as in SVG, CSS, etc. It's one of your bullet points. They are not really colour spaces as those would be CMYK, RGB, rgba, etc.

> * What I'm I expected to do in this PR?

Has nothing to do with the PR.

> * Is `xcolor` package in V3 now?

No, not yet. But I have an implementation of a lot of the conversion methods between different colour spaces already. We can base xcolor on top of your color package implementation. I envision that we will abstract over the colour space class, i.e., have an interface and abstract class for colour ranges, with default named colours that can be extended with definecolor commands. And a set of utilities for colour space transformation.

zorkow commented:
Colour ranges is probably the wrong name. Named colour sets or colour name spaces might be better.
E.g., the current `color` extension implements the dvips names. `xcolor` also has the named colours for svg and x11. CSS names would probably also be useful. I believe there is a css-color style in LaTeX.
Anyway, that has nothing to do with this PR...

OmarIthawi commented on mathjax3-ts/input/tex/color/ColorMethods.ts:
```diff
@@ -0,0 +1,298 @@
+
+
+/**
+ * Override \color macro definition.
```
@zorkow not sure that I understand what that mean.

OmarIthawi replied:
@zorkow I'm not sure that I understand what's needed to be done here.

OmarIthawi commented:
@zorkow @dpvc I've completed all the outstanding comments, rebased over the latest `master` (as of Jan 6th), and created the class `ColorModel` to abstract the color retrieval and definition logic.

Please take a look at it and let me know what do you think.

OmarIthawi commented:
@zorkow A gentle reminder about this PR.

dpvc commented on mathjax3-ts/input/tex/color/ColorMethods.ts:
```diff
@@ -0,0 +1,298 @@
+        delete parser.stack.env['color'];
+    }
+
+    const node = parser.configuration.nodeFactory.create('node', 'mstyle', [math], { mathcolor: color });
```
If I understand correctly, Volker was suggesting you remove the `configuration.nodeFactory` and just use `parser.create()`, but that doesn't seem to have happened.  Is there a reason for that?

dpvc commented on mathjax3-ts/input/tex/color/ColorMethods.ts:
```diff
@@ -0,0 +1,298 @@
+    const cname = parser.GetArgument(name),
+        math = ParseUtil.internalMath(parser, parser.GetArgument(name));
+
+    const node = parser.configuration.nodeFactory.create('node', 'mpadded', math, {
```
Again, same change here.

dpvc commented on mathjax3-ts/input/tex/color/ColorMethods.ts:
```diff
@@ -0,0 +1,298 @@
+        cname = parser.GetArgument(name),
+        math = ParseUtil.internalMath(parser, parser.GetArgument(name));
+
+    const node = parser.configuration.nodeFactory.create('node', 'mpadded', math, {
```
Same again.

dpvc commented on mathjax3-ts/input/tex/color/ColorUtil.ts:
```diff
@@ -0,0 +1,194 @@
+import { COLORS } from './ColorConstants.js';
+
+export type ColorModelProcessor = (def: string) => string;
+export type MethodsMap = Record<string, ColorModelProcessor>;
```
Should this be a `Map` rather than a `Record` (like `COLORS` is?).

OmarIthawi replied:
@dpvc not really. All of `BaseMethods`, `AmsMethods` and `BoldsymbolMethods` are `Record`s. 

Unless we'd like to refactor all of the others, I'd like to make Color.js consistent with them.

dpvc replied:
The thing that corresponds to `BaseMethods`, etc, is `ColorMethods` in `ColoreMethods.ts` above.  The `ColorMethodProcessors` below is specific to your code, and can be handled however you want.  Historically, we've had to use generic objects (`Record`s) for this type of task because Javascript didn't have anything better, and when user input is used as the property key, it could lead to unexpected results, because general objects have lots of pre-defined keys (like `toString` for example).  This leads to the need to use `hasOwnProperty()`, which is awkward.  You do use it here (line 50), so are safe, but now that Javascript has Map objects for just this type of situation, it seems that one should really use them.  I admit that I don't always remember to do that myself, but I'm trying to move over to maps wherever possible.

zorkow commented:
Indeed, sorry for taking so long. As @dpvc mentioned we were busy with travelling and this week are publication deadlines relevant to one of the projects we are currently working on. I will have another look next week.

OmarIthawi commented:
Thanks David! I'm off till next week. I'll address the notes by then. Best
regards!

On Tue, Jan 29, 2019, 18:39 Davide P. Cervone <notifications@github.com
wrote:

> *@dpvc* requested changes on this pull request.
>
> sorry for the delay. Volker and I were just back from the Joint Math
> Meetings when your message came in, and I at least was trying to catch up
> after being away.
>
> I've added a few small comments here. Nothing serious, so I think it is
> ready to go after that.
> ------------------------------
>
> In mathjax3-ts/input/tex/color/ColorMethods.ts
> <https://github.com/mathjax/mathjax-v3/pull/136#discussion_r251874852>:
>
> > +ColorMethods.TextColor = function (parser: TexParser, name: string) {
> +    const model = parser.GetBrackets(name, '');
> +    const colorDef = parser.GetArgument(name);
> +    const color = getColor(parser.options, model, colorDef);
> +    const old = parser.stack.env['color'];
> +
> +    parser.stack.env['color'] = color;
> +    const math = parser.ParseArg(name);
> +
> +    if (old) {
> +        parser.stack.env['color'] = old;
> +    } else {
> +        delete parser.stack.env['color'];
> +    }
> +
> +    const node = parser.configuration.nodeFactory.create('node', 'mstyle', [math], { mathcolor: color });
>
> If I understand correctly, Volker was suggesting you remove the
> configuration.nodeFactory and just use parser.create(), but that doesn't
> seem to have happened. Is there a reason for that?
> ------------------------------
>
> In mathjax3-ts/input/tex/color/ColorMethods.ts
> <https://github.com/mathjax/mathjax-v3/pull/136#discussion_r251875398>:
>
> > +    }
> +
> +    parser.options.definedColors.set(cname, getColor(parser.options, model, def));
> +};
> +
> +/**
> + * Produce a text box with a colored background: `\colorbox`.
> + *
> + * @param {TexParser} parser The calling parser.
> + * @param {string} name The name of the control sequence.
> + */
> +ColorMethods.ColorBox = function (parser: TexParser, name: string) {
> +    const cname = parser.GetArgument(name),
> +        math = ParseUtil.internalMath(parser, parser.GetArgument(name));
> +
> +    const node = parser.configuration.nodeFactory.create('node', 'mpadded', math, {
>
> Again, same change here.
> ------------------------------
>
> In mathjax3-ts/input/tex/color/ColorMethods.ts
> <https://github.com/mathjax/mathjax-v3/pull/136#discussion_r251876618>:
>
> > +    NodeUtil.setProperties(node, padding(parser.options.colorPadding));
> +    parser.Push(node);
> +};
> +
> +/**
> + * Procude a framed text box with a colored background: `\fcolorbox`.
> + *
> + * @param {TexParser} parser The calling parser.
> + * @param {string} name The name of the control sequence.
> + */
> +ColorMethods.FColorBox = function (parser: TexParser, name: string) {
> +    const fname = parser.GetArgument(name),
> +        cname = parser.GetArgument(name),
> +        math = ParseUtil.internalMath(parser, parser.GetArgument(name));
> +
> +    const node = parser.configuration.nodeFactory.create('node', 'mpadded', math, {
>
> Same again.
> ------------------------------
>
> In mathjax3-ts/input/tex/color/ColorUtil.ts
> <https://github.com/mathjax/mathjax-v3/pull/136#discussion_r251880124>:
>
> > + *  See the License for the specific language governing permissions and
> + *  limitations under the License.
> + */
> +
> +/**
> + * @fileoverview Utility functions and classes for the color package.
> + *
> + * @author i@omardo.com (Omar Al-Ithawi)
> + */
> +
> +
> +import TexError from '../TexError.js';
> +import { COLORS } from './ColorConstants.js';
> +
> +export type ColorModelProcessor = (def: string) => string;
> +export type MethodsMap = Record<string, ColorModelProcessor>;
>
> Should this be a Map rather than a Record (like COLORS is?).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mathjax/mathjax-v3/pull/136#pullrequestreview-197605616>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAnYJPO7Dbp_ieQ6otFNr1D1dPw9kBmWks5vIGsWgaJpZM4Vf5Ol>
> .
>

OmarIthawi commented:
Thank you again for getting back promptly @zorkow @dpvc!

I've addressed the minor changes you've suggested. Please take another look at your earliest convenience.

Safe travels!

dpvc commented:
Thanks!  I have approved the PR.  When Volker finishes his review, I think you are ready to go!

Thanks for your work on this, and for your patience with the review process.

OmarIthawi commented:
Great to hear that @dpvc! Looking forward to hearing from @zorkow!

Thank you both for the detailed review.

zorkow commented:
@OmarIthawi Once the remaining small comments are resolved and we merge, I can port and finish my xcolor implementation and inherit from the ColorModel/Processor classes. There might be some issues that come to light then and I will get back to you.
But should go a lot faster now. Again sorry about all the delays.

OmarIthawi commented on mathjax3-ts/input/tex/color/ColorUtil.ts:
```diff
@@ -0,0 +1,193 @@
+            return COLORS.get(name);
+        }
+
+        return name;
```
It does, I'm adding docs to explain why this is needed. See:

https://github.com/mathjax/MathJax/blob/419b0a6eee7eefc0f85e47f7d4f8227ec28b8e57/unpacked/extensions/TeX/color.js#L185-L188

OmarIthawi commented:
Thanks @zorkow, I've addressed all the review comments. Please take another look.

I'm sure some issues will appear when making the `xcolor` plugin.

OmarIthawi commented on mathjax3-ts/input/tex/color/ColorUtil.ts:
```diff
@@ -0,0 +1,193 @@
+    getColorByName(name: string): string {
+        if (this.userColors.has(name)) {
+            return this.userColors.get(name);
+        } else if (COLORS.has(name)) {
```
done

OmarIthawi commented on mathjax3-ts/input/tex/color/ColorUtil.ts:
```diff
@@ -0,0 +1,193 @@
+        return name;
+    }
+
+    public defineColor(model: string, name: string, def: string) {
```
done

OmarIthawi commented on mathjax3-ts/input/tex/color/ColorMethods.ts:
```diff
@@ -0,0 +1,155 @@
+};
+
+/**
+ * Procude a framed text box with a colored background: `\fcolorbox`.
```
good catch. thanks!

zorkow commented:
@OmarIthawi As per face to face with @dpvc we will make this package the default behaviour for the color command and refactor the old functionality into a separate extension package as apparently some people depend on its behaviour.

pkra commented:
Congrats, @OmarIthawi! Awesome job!

OmarIthawi commented:
Thanks @zorkow and @dpvc for the review! I certainly learned a lot!

Thank @pkra! It's always good to hear from you :smiley:

pkra commented:
@OmarIthawi  :eyes:  I'm always watching :wink:
